### DISCUSSION TITLE
Buttons working on iOS + WEB + Desktop WhatsApp, but not for Android

### DISCUSSION THREAD

**Discussion by FreedomLabsIO** [tulir/whatsmeow#650] [WhatsApp protocol Q&A]
Hi! Is there any secret to make a button message arrive on an Android device ?

My code does send a message button to all other 3 devices: iOS, Web and Windows Desktop successfully, but on Android receiver nothing appears on the chat.

It's strange though, because I tried a service that sends an OTP code to WhatsApp numbers (there's a copy button on the message) and the Android device received it.

Is anything wrong with my code?

I even tried to clone the entire message from the OTP example so it's exactly the same, even with that the message does not display.

FYI: I've tried mocking "TemplateID", "RecipientKeyHash", "RecipientKeyIndexes", "RecipientTimestamp" to the same as the message that arrived on the Android device (sent by the OTP service I've tested).

```
func (cli *Client) GetMessageWithButton(title, content, footer, buttonText string) *waE2E.Message {

	var hyderatedButtons []*waE2E.HydratedTemplateButton
	index := uint32(0)

	hyderatedButtons = append(hyderatedButtons, &waE2E.HydratedTemplateButton {
		Index: &index,
		HydratedButton: &waE2E.HydratedTemplateButton_UrlButton {
			UrlButton: &waE2E.HydratedTemplateButton_HydratedURLButton {
				DisplayText: proto.String(buttonText),
				URL: proto.String("https://www.whatsapp.com/otp/code/?otp_type=COPY_CODE&code_expiration_minutes=10&code=otp881320"),
			}}})

	timestamp, _ := strconv.ParseUint("1724818629", 10, 64)

	hydratedTemplate := &waE2E.TemplateMessage_HydratedFourRowTemplate {
		HydratedContentText: proto.String(content),
		HydratedFooterText:  proto.String(footer),
		HydratedButtons:     hyderatedButtons,
		TemplateID:          proto.String("4194019344155670"),
		Title: &waE2E.TemplateMessage_HydratedFourRowTemplate_HydratedTitleText {
			HydratedTitleText: proto.String(title),
		},
	}

	templateMessage := &waE2E.TemplateMessage {
		HydratedTemplate: hydratedTemplate,
		TemplateID: proto.String("4194019344155670"),
	}

	deviceListMetaData := &waE2E.DeviceListMetadata {
		RecipientKeyHash: []byte{}, // []byte{0xba, 0x35, 0xc5, 0x4d, 0xb1, 0x01, 0x3c, 0x62, 0xef, 0x44}
		RecipientTimestamp: &timestamp,
		RecipientKeyIndexes: []uint32{}, // []uint32{0, 15, 2}
	}

	messageContextinfo := &waE2E.MessageContextInfo{
		DeviceListMetadataVersion: proto.Int32(2),
		DeviceListMetadata: deviceListMetaData,
	}

	message := &waE2E.Message {
		TemplateMessage: templateMessage,
		MessageContextInfo: messageContextinfo,
	}

	// msgViewOnce := &waE2E.Message{
	// 	ViewOnceMessage: &waE2E.FutureProofMessage {
	// 		Message: message,
	// }}

	return message
}
```
```
buttonMessageTemplate := cli.GetMessageWithButton("", "Your verification code is *881320*.", "This code expires in 10 minutes.", "COPY CODE")
cli.SendMessage(context.Background(), cli.Store.ID.ToNonAD(), buttonMessageTemplate)
```

**Comment by masewo** [comment]
Since the latest changes from WhatsApp it should only be possible with the official WhatsApp Business API afaik.

**Reply by FreedomLabsIO** [reply]
Yeah, I've seen this answers like these here since 2022, but it does not make sense, since it still works for iOS + Web + Desktop.

Also the message does arrive on the Android device (I can see it in the logs with Frida), it's just not displayed for some reason.

There must be a trick to make it work.

**Reply by masewo** [reply]
Yea, not really logical but I guess they do an extra verification on Android devices. Whiskey stopped supporting it too.

**Comment by a5r0n** [comment]
@brunoaduarte 
About the OTP service, if it's not an official bot, you could try to get the proto message and find any difference or tricks they use.

**Reply by FreedomLabsIO** [reply]
Seems it's more complicated than that. I've debugged the apk and from my tests what I see is that the Android app detects if the message is coming from a companion or from the main device. If it's from a companion device and it has buttons on it, it discards and sends a ack error.

`<ack id='3EB0EA003CAFE84874AAA7' to='xxxxxxxxxxxx.0:63@s.whatsapp.net' class='message' type='text' error='491'/>`

So it only accepts interactive messages from main devices. The only solution would be to use a `yowsup` like sender that mocks a real device. **whatsmeow** or any other companion based apps will not work when sending interactive messages to Android devices.

**Reply by w3nder** [reply]
On android with the
 ```
"quotedMessage": {
      "interactiveMessage": {
        "body": {
          "text": "msg de botao"
        },
        "nativeFlowMessage": {
          "buttons": [.....
```

Can display the buttons.

**Reply by FreedomLabsIO** [reply]
@w3nder , i tried applying your structure to whatsmeow, but it din't worked. The message just appeared as a normal quoted message

```
func (cli *Client) GetButtons(title, content, footer, buttonText string) *waE2E.Message {
	msgVersion := int32(1)
	return &waE2E.Message {
		InteractiveMessage: &waE2E.InteractiveMessage {
			Header: &waE2E.InteractiveMessage_Header {
				Title: proto.String(title),
				Subtitle: proto.String(content),
			},
			Footer: &waE2E.InteractiveMessage_Footer {
				Text: proto.String(footer),
			},
			Body: &waE2E.InteractiveMessage_Body {
				Text: proto.String(buttonText),
			},
			InteractiveMessage: &waE2E.InteractiveMessage_NativeFlowMessage_ {
				NativeFlowMessage: &waE2E.InteractiveMessage_NativeFlowMessage {
					MessageVersion: &msgVersion,
					Buttons: []*waE2E.InteractiveMessage_NativeFlowMessage_NativeFlowButton {
						{
							Name: proto.String("Yes"),
						},
						{
							Name: proto.String("No"),
						},
					},
				},
			},
			ContextInfo: &waE2E.ContextInfo {

			},
		},
	}
}

func (cli *Client) GetQuote(to types.JID, msgToQuote *waE2E.Message) *waE2E.Message {
	return &waE2E.Message {
		ExtendedTextMessage: &waE2E.ExtendedTextMessage {
			Text: proto.String(""),
			ContextInfo: &waE2E.ContextInfo {
				StanzaID: proto.String(""),
				Participant: proto.String(to.ToNonAD().String()),
				QuotedMessage: msgToQuote,
			},
		},
	}
}
```
```
to := types.NewJID("mynumber_xxxxxx", types.DefaultUserServer)
cli.SendMessage(context.Background(), to, cli.GetButtons("Title", "SubTitle", "Footer", "msg de botao"))
cli.SendMessage(context.Background(), to, cli.GetQuote(to, cli.GetButtons("Title", "SubTitle", "Footer", "msg de botao")))
```
And the result is
![image](https://github.com/user-attachments/assets/b12b951b-b6c8-45f8-b02d-cae52d1a870e)

Any ideas?

**Reply by w3nder** [reply]
We are able to send the buttons to Baileys in the way I mentioned.

Let me see how it looks in Meow.

**Reply by Ased2235** [reply]
> > We are able to send the buttons to Baileys in the way I mentioned.
> > Let me see how it looks in Meow.
> 
> Nevermind, it worked! Thanks!

Hello, Mind posting the full code which worked for you?

**Reply by badlee** [reply]
> > We are able to send the buttons to Baileys in the way I mentioned.
> > Let me see how it looks in Meow.
> 
> Nevermind, it worked! Thanks!

@brunoaduarte, which code worked for you?

**Reply by FreedomLabsIO** [reply]
@w3nder the buttons doesn't work for iOS, nothing is received on the device. Is it working for you?

```
type InteractiveMessage_NativeFlowMessage_NativeFlowButton struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	Name             *string `protobuf:"bytes,1,opt,name=name" json:"name,omitempty"`
	ButtonParamsJSON *string `protobuf:"bytes,2,opt,name=buttonParamsJSON" json:"buttonParamsJSON,omitempty"`
}
```

```

func (cli *Client) GetButtons(title, content, footer string, buttonText []string) *waE2E.Message {

	var nativeFlowButtons []*waE2E.InteractiveMessage_NativeFlowMessage_NativeFlowButton

	msgVersion := int32(1)
	deviceListMetadataVersion := int32(2)

        for _, button := range buttonText {

	            jsonData, _ := json.Marshal(ButtonParams{
		            DisplayText: button,
		            ID:          ".menu",
	            })
            
	            nativeFlowButtons = append(nativeFlowButtons, &waE2E.InteractiveMessage_NativeFlowMessage_NativeFlowButton{
		            Name:             proto.String("quick_reply"),
		            ButtonParamsJSON: proto.String(string(jsonData)),
	            })

        }

	return &waE2E.Message{
		InteractiveMessage: &waE2E.InteractiveMessage{
			Header: &waE2E.InteractiveMessage_Header{
				Title:              proto.String(title),
				Subtitle:           proto.String(content),
				HasMediaAttachment: proto.Bool(false),
			},
			Footer: &waE2E.InteractiveMessage_Footer{
				Text: proto.String(footer),
			},
			Body: &waE2E.InteractiveMessage_Body{
				Text: proto.String(content),
			},
			InteractiveMessage: &waE2E.InteractiveMessage_NativeFlowMessage_{
				NativeFlowMessage: &waE2E.InteractiveMessage_NativeFlowMessage{
					MessageVersion: &msgVersion,
					Buttons:        nativeFlowButtons,
				},
			},
			ContextInfo: &waE2E.ContextInfo{},
		},
		MessageContextInfo: &waE2E.MessageContextInfo{
			DeviceListMetadata:        &waE2E.DeviceListMetadata{},
			DeviceListMetadataVersion: &deviceListMetadataVersion,
		},
	}
}

cli.GetButtons("title", "content", "footer", []string{"button"}, nil)
```

BTW, if I comment
```
// Name:             proto.String("quick_reply"),
```

The message is received, but the button is empty

![image](https://github.com/user-attachments/assets/e96d127e-b151-4aaf-a2d2-8124c62d0f8d)

**Comment by wxl2578** [comment]
Is there currently a code for the send button? In addition, when sending a template message, what is the message type and media type?

**Reply by wxl2578** [reply]
media  image

**Comment by hrizal** [comment]
at this time you cannot do that, i think meta filtered messages like that
if coming from regular whatsapp.

Pada Sel, 8 Okt 2024 pukul 20.01 Patricia C Stryker <
***@***.***> menulis:

**Comment by f-amine** [comment]
any updates about this ? it works fine for android but not IOS for me

**Comment by hrizal** [comment]
No one its work since 2023

Pada Kam, 14 Nov 2024 10.47, Bruno Duarte ***@***.***>
menulis:

**Comment by JoaoArtur** [comment]
Hello, anyone using in 2025?

**Reply by Adamkaram** [reply]
does anything work for now?
